### PR TITLE
IGNITE-23063 ItReplicaLifecycleTest.testStableAreWrittenAfterRestart(TestInfo) is flaky

### PR DESCRIPTION
**Pull request by JAkutenshi**
JIRA Ticket: [IGNITE-23063 | ItReplicaLifecycleTest.testStableAreWrittenAfterRestart(TestInfo) is flaky](https://issues.apache.org/jira/browse/IGNITE-23063)

## The goal

The main goal is to fix the mentioned tests and other flaky tests in `ItReplicaLifecycleTest`.

## The reason

There was an `AssetionError` with a message "Raft nodes <...> are still running on the Ignite node" inside `Loza#stopAsync`'s assertion `assert nodes.isEmpty()`. The reason may be found in logs: there is a scenario:
1. Ignite node is stopping
2. Inside of zone replica starting, at the beginning there no busy lock and then raft node will be started.
3. Then, there will be starting replica process and it will be interrupted because of busy lock, thus there no the corresponding replica's future in `replicas` map.
4. Then, while `replicaManager#beforeNodeStop` there will not be any stopped raft node because there are no the corresponding zone replica future that owes the reference on raft node. 
  - Also, before the fix, there was `weakStopReplica` that shouldn't be used on Ignite node's stopping.
6. In the end, when raft manager (Loza) will be stopping, there are unstopped lost raft nodes.

## The solution

1. Revert `weakStopReplica` to `Replica#stopReplica` inside of `ReplicaManager#stopPartition` inside of `ReplicaManager#cleanUpPartitionsResources`.
2. Extract `internalZoneReplicaStart` and call it only after busy lock acquiring.
3. Add raft-node with corresponding zone replication group identifier stopping if we couldn't acquire busy lock inside start zone replica's starting future stage.

---

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 1]
```diff
@@ -709,6 +737,12 @@ public CompletableFuture<Replica> startReplica(
 
         return newRaftClientFut.thenComposeAsync(raftClient -> {
             if (!busyLock.enterBusy()) {
+                try {
+                    raftManager.stopRaftNodes(replicaGrpId);
```
Comment (kgusakov) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:741]: Looks strange, why we stop raft node on start? If without this code everything is still broken, I think we should to fix the real reason of race

**Review by kgusakov — CHANGES_REQUESTED** [review]

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 2]
```diff
@@ -1246,7 +1280,7 @@ CompletableFuture<Boolean> weakStartReplica(
             synchronized (context) {
                 ReplicaState state = context.replicaState;
 
-                LOG.debug("Weak replica start [grp={}, state={}, future={}].", groupId, state, context.previousOperationFuture);
+                LOG.info("Weak replica start [grp={}, state={}, future={}].", groupId, state, context.previousOperationFuture);
```
[the remark this answers is not in the document]
Comment (JAkutenshi) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1283]: Oh, its a mistake

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 3]
```diff
@@ -926,6 +959,23 @@ public CompletableFuture<Void> stopAsync(ComponentContext componentContext) {
         shutdownAndAwaitTermination(executor, shutdownTimeoutSeconds, TimeUnit.SECONDS);
         shutdownAndAwaitTermination(replicasCreationExecutor, shutdownTimeoutSeconds, TimeUnit.SECONDS);
 
+        try {
+            CompletableFuture.allOf(raftClientsFutures.values().toArray(new CompletableFuture<?>[0]))
+                    .get(shutdownTimeoutSeconds, TimeUnit.SECONDS);
+
+            raftClientsFutures.forEach((raftNodeId, raftClientFuture) -> {
+                raftClientFuture.thenAccept(RaftGroupService::shutdown);
```
Comment (kgusakov) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:967]: Any exceptions from shutdown will be ignored

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 4]
```diff
@@ -926,6 +959,23 @@ public CompletableFuture<Void> stopAsync(ComponentContext componentContext) {
...
+
+                try {
+                    raftManager.stopRaftNode(raftNodeId);
```
Comment (kgusakov) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:970]: Exception on the one raftNodeId will stop processing of the whole chain

**Review by kgusakov — CHANGES_REQUESTED** [review]

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 5]
```diff
@@ -926,6 +959,23 @@ public CompletableFuture<Void> stopAsync(ComponentContext componentContext) {
...
+                } catch (NodeStoppingException e) {
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:971]: I don't think that it's possible to catch following exception here. Could you please elaborate a scenario?

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 6]
```diff
@@ -926,6 +959,23 @@ public CompletableFuture<Void> stopAsync(ComponentContext componentContext) {
...
+                    LOG.debug("Raft node is already stopping [raftNodeId={}]", raftNodeId, e);
+                }
+            });
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:975]: Please use graceful stop, like org.apache.ignite.internal.util.IgniteUtils#closeAll(java.util.stream.Stream<? extends java.lang.AutoCloseable>)

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 3 above)
Comment (JAkutenshi) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:967]: Fixed with `AssertionError`

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 4 above)
Comment (JAkutenshi) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:970]: Fixed with `AssertionError`

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 5 above)
Comment (JAkutenshi) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:971]: Fixed with a proper `AssertionError` with a message why it's an abnormal case

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 1 above)
Comment (JAkutenshi) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: The implementation is changed in way to collect raft-nodes and stopping them manually on `ReplicaManager#stopAsync` if there are unreachable nodes from replicas.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 3 above)
Comment (JAkutenshi) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: Raft client shutdown process is removed now in favor of https://issues.apache.org/jira/browse/IGNITE-23109

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 6 above)
Comment (JAkutenshi) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:974]: Raft client shutdown process is removed now in favor of https://issues.apache.org/jira/browse/IGNITE-23109

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 7]
```diff
@@ -926,14 +960,28 @@ public CompletableFuture<Void> stopAsync(ComponentContext componentContext) {
...
+                            raftManager.stopRaftNode(raftNodeId);
+                        } catch (NodeStoppingException e) {
+                            throw new AssertionError("Raft node is stopping [raftNodeId=" + raftNodeId
+                                    + "], but it's abnormal, because Raft Manager must stop strictly after Replica Manager.", e);
+                        }
+                    }),
+                    () -> replicas.values().forEach(replicaFuture -> replicaFuture.completeExceptionally(new NodeStoppingException()))
+            );
+        } catch (Exception e) {
+            LOG.error("There are exceptions during Replica Manager stop", e);
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:978]: It's stopAsync, thus it's better to return failedFuture(e);

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 8]
```diff
@@ -926,14 +960,28 @@ public CompletableFuture<Void> stopAsync(ComponentContext componentContext) {
         shutdownAndAwaitTermination(executor, shutdownTimeoutSeconds, TimeUnit.SECONDS);
         shutdownAndAwaitTermination(replicasCreationExecutor, shutdownTimeoutSeconds, TimeUnit.SECONDS);
 
+        try {
+            IgniteUtils.closeAllManually(
+                    () -> CompletableFuture.allOf(raftClientsFutures.values().toArray(new CompletableFuture<?>[0]))
+                            .get(shutdownTimeoutSeconds, TimeUnit.SECONDS),
+                    () -> raftClientsFutures.forEach((raftNodeId, raftClientFuture) -> {
+                        try {
+                            raftManager.stopRaftNode(raftNodeId);
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:969]: Am I right that you've wrapped all raftNodes stop into single close?

Comment (JAkutenshi) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:969]: Yes, I see, should be separated

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 7 above)
Comment (JAkutenshi) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: Done!

**Review by sanpwc — APPROVED** [review]

**Review by kgusakov — APPROVED** [review]

**Pull request merged**